docs/code/client_middleware_cookbook.py

```diff
@@ -0,0 +1,132 @@
+"""This is a collection of semi-complete examples that get included into the cookbook page."""
+
+import asyncio
+import logging
+import time
+from collections.abc import AsyncIterator, Sequence
+from contextlib import asynccontextmanager, suppress
+
+from aiohttp import (
+    ClientHandlerType,
+    ClientRequest,
+    ClientResponse,
+    ClientSession,
+    TCPConnector,
+)
+from aiohttp.abc import ResolveResult
+from aiohttp.tracing import Trace
+
+
+class SSRFError(Exception):
+    """A request was made to a blacklisted host."""
+
+
+async def retry_middleware(
+    req: ClientRequest, handler: ClientHandlerType
+) -> ClientResponse:
+    for _ in range(3):  # Try upto 3 times
+        resp = await handler(req)
+        if resp.ok:
+            return resp
+    assert False  # Unreachable
+
+
+async def api_logging_middleware(
+    req: ClientRequest, handler: ClientHandlerType
+) -> ClientResponse:
+    # We use middlewares=() to avoid infinite recursion.
+    async with req.session.post("/log", data=req.url.host, middlewares=()) as resp:
+        if not resp.ok:
+            logging.warning("Log endpoint failed")
+
+    return await handler(req)
+
+
+class TokenRefresh401Middleware:
+    def __init__(self, refresh_token: str, access_token: str):
+        self.access_token = access_token
+        self.refresh_token = refresh_token
+        self.lock = asyncio.Lock()
+
+    async def __call__(
+        self, req: ClientRequest, handler: ClientHandlerType
+    ) -> ClientResponse:
+        for _ in range(2):  # Retry at most one time
+            token = self.access_token
+            req.headers["Authorization"] = f"Bearer {token}"
+            resp = await handler(req)
+            if resp.status != 401:
+                return resp
+            async with self.lock:
+                if token != self.access_token:  # Already refreshed
+                    continue
+                url = "https://api.example/refresh"
+                async with req.session.post(url, data=self.refresh_token) as resp:
+                    self.access_token = await resp.json()
```

Comment:
This looks a bit strange to store the whole json payload as the token

Reply:
I was just keeping it simple, feel free to make it more realistic.

Reply:
bace3a9a9